java/common/src/main/java/org/apache/shindig/common/servlet/URLFilter.java

```diff
@@ -0,0 +1,127 @@
+    private static Logger log = Logger.getLogger(URLFilter.class.getName());
+
+    private List allowedURLs;
+
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+
+        if (httpRequest.getMethod().equalsIgnoreCase("POST")) {
+
+            // MultiReadHttpServletRequest is used to read request data more than once.
+            MultiReadHttpServletRequest multiReadRequest = new MultiReadHttpServletRequest(httpRequest);
+
+            if (handlePOSTRequest(multiReadRequest)) {
+                chain.doFilter(multiReadRequest, response);
+            } else {
+                HttpServletResponse httpResponse = (HttpServletResponse) response;
+                httpResponse.sendError(HttpServletResponse.SC_FORBIDDEN, "Unauthorized body parameter detected!");
+            }
+
+        } else if (httpRequest.getMethod().equalsIgnoreCase("GET")) {
+
+            if (!isInvalidHostPresent(URLDecoder.decode(httpRequest.getQueryString(), StandardCharsets.UTF_8.name()))) {
+                chain.doFilter(httpRequest, response);
+            } else {
+                HttpServletResponse httpResponse = (HttpServletResponse) response;
+                httpResponse.sendError(HttpServletResponse.SC_FORBIDDEN, "Unauthorized query parameter detected!");
+            }
+        }
+    }
+
+    private boolean handlePOSTRequest(MultiReadHttpServletRequest multiReadRequest) {
+
+        StringBuilder stringBuffer = new StringBuilder();
+        String line;
+        try {
+            BufferedReader reader = multiReadRequest.getReader();
+            while ((line = reader.readLine()) != null) {
+                stringBuffer.append(line);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return !isInvalidHostPresent(stringBuffer.toString());
+    }
+
+    private boolean isInvalidHostPresent(String text) {
+
+        Pattern pattern = Pattern.compile(URL_REGEX);
+        Matcher matcher = pattern.matcher(text);
+
+        while (matcher.find()) {
+            String match = matcher.group();
+            try {
+                URI uri = new URI(match);
+                if (uri.getHost() != null && !allowedURLs.contains(uri.getHost())) {
+                    log.warning("Potential External Service Interaction (DNS) attack thwarted. Unauthorized " +
+                            "host name: " + uri.getHost() + " detected in shindig web app request parameters.");
+                    return true;
+                }
+            } catch (URISyntaxException e) {
+                e.printStackTrace();
+            }
+        }
+        return false;
+    }
+
+    public void destroy() {
+    }
+
+    public void init(FilterConfig filterConfig) {
+
+        String allowedURLString = filterConfig.getInitParameter(ALLOWED_URLS_PARAM_NAME);
```

Comment:
In the mail we also decided to get this allowedURLString from a system param if the filter init param was unavailable. Shall we add that part as well?